homeassistant/components/crownstone/light.py

```diff
@@ -0,0 +1,204 @@
+"""Support for Crownstone devices."""
+from __future__ import annotations
+
+from collections.abc import Mapping
+from functools import partial
+import logging
+from typing import TYPE_CHECKING, Any
+
+from crownstone_cloud.cloud_models.crownstones import Crownstone
+from crownstone_cloud.const import (
+    DIMMING_ABILITY,
+    SWITCHCRAFT_ABILITY,
+    TAP_TO_TOGGLE_ABILITY,
+)
+from crownstone_cloud.exceptions import CrownstoneAbilityError
+from crownstone_uart import CrownstoneUart
+
+from homeassistant.components.light import (
+    ATTR_BRIGHTNESS,
+    SUPPORT_BRIGHTNESS,
+    LightEntity,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import (
+    ABILITY_STATE,
+    CROWNSTONE_INCLUDE_TYPES,
+    CROWNSTONE_SUFFIX,
+    DOMAIN,
+    SIG_CROWNSTONE_STATE_UPDATE,
+    SIG_UART_STATE_CHANGE,
+)
+from .devices import CrownstoneDevice
+from .helpers import map_from_to
+
+if TYPE_CHECKING:
+    from .entry_manager import CrownstoneEntryManager
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up crownstones from a config entry."""
+    manager: CrownstoneEntryManager = hass.data[DOMAIN][config_entry.entry_id]
+
+    entities: list[CrownstoneEntity] = []
+
+    # Add Crownstone entities that support switching/dimming
+    for sphere in manager.cloud.cloud_data:
+        for crownstone in sphere.crownstones:
+            if crownstone.type in CROWNSTONE_INCLUDE_TYPES:
+                # Crownstone can communicate with Crownstone USB
+                if manager.uart and sphere.cloud_id == manager.usb_sphere_id:
+                    entities.append(CrownstoneEntity(crownstone, manager.uart))
+                # Crownstone can't communicate with Crownstone USB
+                else:
+                    entities.append(CrownstoneEntity(crownstone))
+
+    async_add_entities(entities)
+
+
+def crownstone_state_to_hass(value: int) -> int:
+    """Crownstone 0..100 to hass 0..255."""
+    return map_from_to(value, 0, 100, 0, 255)
+
+
+def hass_to_crownstone_state(value: int) -> int:
+    """Hass 0..255 to Crownstone 0..100."""
+    return map_from_to(value, 0, 255, 0, 100)
+
+
+class CrownstoneEntity(CrownstoneDevice, LightEntity):
+    """
+    Representation of a crownstone.
+
+    Light platform is used to support dimming.
+    """
+
+    _attr_should_poll = False
+    _attr_icon = "mdi:power-socket-de"
+
+    def __init__(self, crownstone_data: Crownstone, usb: CrownstoneUart = None) -> None:
```

Comment:
`usb` should be typed as optional.

Reply:
It does not want to work with this property however;
```python
 @property
 def usb_available(self) -> bool:
     """Return if this entity can use a usb dongle."""
     return self.usb is not None and self.usb.is_ready()
```
```python
if self.usb_available:
```

```console
$ mypy --strict homeassistant/components/crownstone
homeassistant/components/crownstone/light.py:162: error: Item "None" of "Optional[Any]" has no attribute "dim_crownstone"  [union-attr]
homeassistant/components/crownstone/light.py:182: error: Item "None" of "Optional[Any]" has no attribute "switch_crownstone"  [union-attr]
homeassistant/components/crownstone/light.py:196: error: Item "None" of "Optional[Any]" has no attribute "switch_crownstone"  [union-attr]
```

Reply:
We probably need to check, assert or cast the type in the relevant place before accessing attributes of the name.